yapapi/network.py

```diff
@@ -0,0 +1,230 @@
+                }
+            ]
+        }
+        return deploy_args
+
+
+class Network:
+    """
+    Describes a VPN created between the requestor and the provider nodes within Golem Network.
+    """
+
+    @classmethod
+    async def create(
+        cls,
+        net_api: "Net",
+        ip: str,
+        owner_id: str,
+        owner_ip: Optional[str] = None,
+        mask: Optional[str] = None,
+        gateway: Optional[str] = None,
+    ):
+        """Create a new VPN.
+
+        :param net_api: the mid-level binding used directly to perform calls to the REST API.
+        :param ip: the IP address of the network. May contain netmask, e.g. "192.168.0.0/24"
+        :param owner_id: the node ID of the owner of this VPN (the requestor)
+        :param owner_ip: the desired IP address of the requestor node within the newly-created Network
+        :param mask: Optional netmask (only if not provided within the `ip` argument)
+        :param gateway: Optional gateway address for the network
+
+        :return: a Network object allowing further manipulation of the created VPN
+        """
+
+        network = cls(net_api, ip, owner_id, owner_ip, mask, gateway)
+
+        # create the network in yagna and set the id
+        network._network_id = await net_api.create_network(
+            network.network_address, network.netmask, network.gateway
+        )
+
+        # add requestor's own address to the network
+        await network.add_owner_address(network.owner_ip)
+
+        return network
+
+    def __init__(
+        self,
+        net_api: "Net",
+        ip: str,
+        owner_id: str,
+        owner_ip: Optional[str] = None,
+        mask: Optional[str] = None,
+        gateway: Optional[str] = None,
+    ):
+        """
+        :param net_api: the mid-level binding used directly to perform calls to the REST API.
+        :param ip: the IP address of the network. May contain netmask, e.g. "192.168.0.0/24"
+        :param owner_id: the node ID of the owner of this VPN (the requestor)
+        :param owner_ip: the desired IP address of the requestor node within the newly-created Network
+        :param mask: Optional netmask (only if not provided within the `ip` argument)
+        :param gateway: Optional gateway address for the network
+        """
+
+        self._net_api = net_api
+        network_ip = f"{ip}/{mask}" if mask else ip
+        try:
+            self._ip_network: IpNetwork = ip_network(network_ip, strict=False)
+        except ValueError as e:
+            raise NetworkError(f"{e}.")
+        self._hosts = self._ip_network.hosts()
+
+        self._network_id: Optional[str] = None
+        self._gateway = gateway
+        self._owner_id = owner_id
+        self._owner_ip: IpAddress = ip_address(owner_ip) if owner_ip else self._next_address()
+
+        self._nodes: Dict[str, Node] = dict()
+        """the mapping between a Golem node id and a Node in this VPN."""
+
+        self._nodes_lock = asyncio.Lock()
```

Comment:
I'm not sure, but I think this lock is redundant now?

There are no `await`s in the locked sections (in both places where it is used), so this code will either way execute as a single part.

The important part that changed since the previous version is 
```
self._nodes[node_id] = node
```
being moved **before** the only await in this function (`await self._net_api.add_node(...)`).

In other words, as long as we added a `node` to `self._nodes` after the `await`, the lock was necessary. Now it's not.

Reply:
I'd leave it be so that we don't need to think about this if, e.g. a future change adds some async code inside